keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,576 @@
+
+
+## Implementation
+
+### CSI changes
+
+Container Storage Interface (CSI) specification will be modified to provide volume health check leveraging existing RPCs and adding new ones.
+
+- GetVolume RPC
+  - External controller calls a new `GetVolume` RPC to check health condition of a particular PV instead of calling `ListVolumes` with a `volume_id` filter. Some CSI drivers chose not to implement `ListVolumes` as it is an expensive operation. Therefore it is better to introduce a new RPC `GetVolume` which has a new capability.
+- NodeGetVolumeStats RPC
+  - For any PV that is mounted, the external agent calls `NodeGetVolumeStats` to see if volume is still mounted;
+  - Calls `NodeGetVolumeStats` to check if volume is usable, e.g., filesystem corruption, bad blocks, etc.
+
+Two new controller capabilities are added. If a CSI driver supports `GET_VOLUME` capability, it MUST support calling `GetVolume` to provide general volume information in `GetVolumeResponse`. If a driver supports `GET_VOLUME_HEALTH`, it MUST provide additional health information in `GetVolumeResponse`.
+
+A new node capability is added. If CSI driver supports the `GET_VOLUME_STATS_HEALTH` capability, it MUST provide health information in `NodeGetVolumeStats`.
+
+Detailed changes needed in the CSI Spec are described in the following.
+
+#### Add GetVolume RPC
+
+Add a new `GetVolume` RPC. `GetVolumeResponse` should contain current information of a volume if it exists. If the volume does not exist any more, `GetVolume` should return gRPC error code `NOT_FOUND`.
+
+Add `VolumeHealth` message in `VolumeStatus` field in `GetVolumeResponse`. This can indicate additional health information for the volume.
+
+Whether the volume is still attached to a node is already handled by `LIST_VOLUMES_PUBLISHED_NODES` controller capability. This capability will be used by Kubernetes to reconcile - re-attach the volume if the volume is not attached any more.
+
+```
+message GetVolumeRequest {
+  // Identity information for a specific volume. This field is
+  // REQUIRED. GetVolume will return with current volume information.
+  string volume_id = 1;
+}
+```
+
+```
+message GetVolumeResponse {
+  message VolumeStatus{
+    // health shows error conditions reported by the SP.
+    // This field MUST be specified if the
+    // GET_VOLUME_HEALTH controller capability is supported.
+    VolumeHealth health = 1;
+  }
+
+  // This field is REQUIRED
+  Volume volume = 1;
+
+  // This field is OPTIONAL. This field MUST be specified if the
+  // GET_VOLUME_HEALTH controller capability is supported.
+  VolumeStatus status = 2;
+}
+```
+
+```
+message VolumeHealth {
+  // The error code describing the health condition of the volume.
+  // This field is REQUIRED.
+  string error_code = 1;
+
+  // The error message associated with the above error_code. This field is OPTIONAL.
+  string message = 2;
+}
+```
+
+The following common error codes are proposed for volume health:
+* VolumeNotFound
```

Comment:
Do you see a reason to separate out specific error codes for Controller vs Node?

Reply:
I think some can be separated but others may be in both category.  Can we sort this out later during the CSI spec review?

Reply:
I removed "NodeFailure" as that will be detected by the external monitoring controller without issuing a CSI call.

Reply:
Yeah we can discuss the error codes in more detail during CSI spec review